:book: plugins doc (#2115)
kubernetes-sigs/kubebuilder: docs/book/src/plugins/creating-plugins.md

# Creating your own plugins

## Overview

> You can extend KubeBuilder API to create your plugins. If you are using the [its extension][extending-cli] then, it means that your plugin will be implemented in your project and register then to the CLI as it has been done by the [SDK][sdk] project. See its [cli code][sdk-cli-pkg] as an example.

Reviewer: ```suggestion
You can extend the Kubebuilder API to create your own plugins. If [extending the CLI][extending-cli], your plugin will be implemented in your project and registered to the CLI as has been done by the [SDK][sdk] project. See its [cli code][sdk-cli-pkg] as an example.
```

## Language-based Plugins

> KubeBuilder offers the Golang-based operator plugins, which will help its CLI tool users create projects following the [Operator Pattern][operator-pattern] Golang language.

Reviewer: ```suggestion
Kubebuilder offers the Golang-based operator plugins, which will help its CLI tool users create projects following the [Operator Pattern][operator-pattern].
```

> Now see that [SDK][sdk], for example, has its language plugins for [Ansible][sdk-ansible] and [Helm][sdk-helm], which are similar options but for users who would like to work with these respective languages and stacks instead of Golang.

Reviewer: ```suggestion
The [SDK][sdk] project, for example, has language plugins for [Ansible][sdk-ansible] and [Helm][sdk-helm], which are similar options but for users who would like to work with these respective languages and stacks instead of Golang.
```

> Note that KubeBuilder provides the `kustomize.common.kubebuilder.io` to help in these efforts. This plugin will scaffold the common base without any specific language scaffold file to allow you to extend the Kubebuilder style for your plugins.

Author: solved

> Then, note that KubeBuilder provides as well the `config-base.kubebuilder.io` to help you with. This plugin will scaffold its common base without any specific language scaffold file to allow you to follow up on the KubeBuilder style for your plugins.  

Reviewer: ```suggestion
Note that KubeBuilder provides the `config-base.kubebuilder.io` to help in these efforts. This plugin will scaffold the common base without any specific language scaffold file to allow you to extend the Kubebuilder style for your plugins.
```

> In this way, currently, you can [Extending the CLI][extending-cli] and use the `Bundle Plugin` to create your language plugins such as:

Reviewer: ```suggestion
In this way, currently, you can [Extend the CLI][extending-cli] and use the `Bundle Plugin` to create your language plugins such as:
```

```go
  mylanguagev1Bundle, _ := plugin.NewBundle(language.DefaultNameQualifier, plugin.Version{Number: 1},
		configbasev1.Plugin{}, // extend the common base from Kuebebuilder
		mylanguagev1.Plugin{}, // your plugin language which will do the scaffolds for the specific language on top of the common base
	)
```

> In case you do not want to develop your using Golang, you can follow up its standard by using its binary such as:

Reviewer: ```suggestion
If you do not want to develop your plugin using Golang, you can follow its standard by using the binary as follows:
```

```sh
kubebuilder init --plugins=config-base/v1
``` 

Then, you can create your implementations for the sub-commands `create api` and `create webhook`, for example, using your language of preference.

<aside class="note">

> <h1>Why follow up the KubeBuilder style?</h1>

Reviewer: ```suggestion
<h1>Why use the Kubebuilder style?</h1>
```

> Note that KubeBuilder and SDK are both projects very adopted which leverages in the [controller-runtime][controller-runtime] project to allow users to build solutions using the [Operator Pattern][operator-pattern] and then, that both are integrated and following up its standards.

Reviewer: ```suggestion
Kubebuilder and SDK are both broadly adopted projects which leverage the [controller-runtime][controller-runtime] project.  They both allow users to build solutions using the [Operator Pattern][operator-pattern] and follow common standards.
```

> Adopting these standards can bring significant benefits, such as joining forces on maintaining KubeBuilder and your solutions by taking advantage of the contributions made to this project by the community as keep your focus to address specific needs please you. 

Reviewer: ```suggestion
Adopting these standards can bring significant benefits, such as joining forces on maintaining Kubebuilder and the common standards maintained there.  This allows you to focus on the specific needs and requirements for your plugin and use-case.
```

> And then, you will also be able to use custom plugins and options currently or in the future provided by these projects as any other which decides to persuade the same standards.

Reviewer: Unsure what we're saying here.  Have we already covered this idea in the previous two sentences?  If so, perhaps we should remove this?

Author: I want to make clear that those who decided to adopt this standard will able not to;  take advantage of what is provided by KB/SDK now and in the future. Also, this person can use any other solution that might came to exist as well which is respecting that.

Author: I will close that as addressed however, please feel free to re-open if you see that the text is not clear. Also, please feel free to raise suggestions for a better re-phase too.

</aside>

## Custom Plugins 

> Note that users are also able to use plugins to customize their scaffold and address specific needs. See that KubeBuilder provides the [declarative][declarative] plugin which can be used when for example an API is scaffold:

Reviewer: ```suggestion
Note that users are also able to use plugins to customize their scaffold and address specific needs. See that Kubebuilder provides the [declarative][declarative-code] plugin which can be used when for example an API is scaffold:
```
Is this the intended link?
